### PR TITLE
Improving Chat design: more muted states design, show topic at the beginning of the history 

### LINKED ISSUE #1763: Show description of chat at beginning of history
When joining a fresh chat, at the beginning of history we want to show the chat description prominently in the timeline.

- [ ] Ideally the admin can also quickly edit it from right there
- [ ] and invite new users.

Additionally we want to 
- [ ] track whether a user already entered a chat room
- [ ] and if they haven't yet and there is a description, we want to prominently show the description once

### PR DESCRIPTION
**Pull request by gnunicorn**
This improves the Chat room in several manners. Commits are atomic about one aspect at a time, so that might be an easier way to review than looking at the entire changeset at a time.

- Show Encryption info message and chat room topic at the beginning of the room history, fixes #1763 . Includes Editing facilities
- Upgrade of the matrix-rust-sdk to provide more robust back-pagination facilities
- Fixes in the state management of back pagination as that wasn't always happen properly, in particular if the room had very few messages
- Refactoring of the chat room message widget into separate files
- New facilities to filter the messages to show already on the provider side, fixing #1662 .
- Several improvements on rendering the different custom messages:
  - membership changes: Show the display name or "you" if it was you; use L10n rather than hard-coded strings; Use a more muted style to be less pungent
  - A new format for encrypted messages with a new tap-for-more-drawer that we can allow for quicker action buttons in the future (e.g. verify device, verify with other user, provide encryption key)

In the real world this looks like the following now:

https://github.com/acterglobal/a3/assets/40496/1c3caa35-8f65-4d08-bf8e-a140250caae7

### CODE REVIEW

File: app/lib/features/chat/utils.dart [hunk 1]
```diff
@@ -6,12 +6,58 @@ import 'package:flutter/material.dart';
...
+      // no support yet
+      return false;
+    case 'm.sticker':
+    case 'm.room.redaction':
+    case 'm.room.encrypted':
+      // supported
+      return true;
+    case 'm.room.message':
+      // special case of supporting locations
+      return message.metadata?['msgType'] == 'm.location';
```
Comment (kumarpalsinh25) [inline app/lib/features/chat/utils.dart:60]: Not getting understanding of this condition.

```
case 'm.room.message':
      // special case of supporting locations
      return message.metadata?['msgType'] == 'm.location';
```

File: app/lib/features/chat/widgets/messages/location_message.dart [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_chat_types/flutter_chat_types.dart';
+
+class LocationMessageWidget extends StatelessWidget {
```
Comment (kumarpalsinh25) [inline app/lib/features/chat/widgets/messages/location_message.dart:4]: Ahh, I see. We have support of the location message as well. I never used/see this in acter.

File: app/lib/features/chat/providers/notifiers/chat_room_notifier.dart [hunk 3]
```diff
@@ -58,7 +58,7 @@ class ChatRoomNotifier extends StateNotifier<ChatRoomState> {
         state = state.copyWith(
           loading: const ChatRoomLoadingState.loading(),
         );
-        final hasMore = await timeline.paginateBackwards(20);
+        final hasMore = !await timeline.paginateBackwards(20);
```
Comment (kumarpalsinh25) [inline app/lib/features/chat/providers/notifiers/chat_room_notifier.dart:61]: Not getting idea about this condition changes.

**Review by kumarpalsinh25 — APPROVED** [review]
Niceee Improvement!!

Added few minor queries but flutter code looks good to me.

File: app/lib/features/chat/utils.dart (hunk 1 above)
Comment (gnunicorn) [reply app/lib/features/chat/utils.dart:60]: it's what we had before, just slightly refactored. I can't tell you the _meaning_ of each of them, just tried to not break them.

File: app/lib/features/chat/widgets/messages/location_message.dart (hunk 2 above)
Comment (gnunicorn) [reply app/lib/features/chat/widgets/messages/location_message.dart:4]: yeah, dunno if we ever implemented sending it...

File: app/lib/features/chat/providers/notifiers/chat_room_notifier.dart (hunk 3 above)
Comment (gnunicorn) [reply app/lib/features/chat/providers/notifiers/chat_room_notifier.dart:61]: I've upgraded the SDK to simplify the code, as a result the API has been updated to not indicate that there is more but that it reached the end. see
https://github.com/acterglobal/a3/pull/1773/files#diff-12c0480bc493ffbb94ce49b75db168b6aec7504019bcbdea587c1192ffe429c2L1128

As a result the state assumptions in the provider for `hasMore` are = `! reachedEnd`. This part of the same commit, it changed together.

**Pull request merged**